### PR TITLE
Makes SyntheticHistory ROM models clusterable

### PR DESCRIPTION
**Pull request by j-bryan**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
#1889 

##### What are the significant changes in functionality due to this change request?
SyntheticHistory ROMs can be used with Clustered and Interpolated ROMCollections to cluster and interpolate models. So far, this capability has been added only for the Fourier and ARMA TSA models.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by j-bryan** [comment]
@dylanjm @PaulTalbot-INL I'd appreciate any feedback you have on this. In the meantime, I'll look at why the precheck is failing.

File: ravenframework/SupervisedLearning/ARMA.py [hunk 1]
```diff
@@ -2527,6 +2528,7 @@ def setEngine(self,eng,seed=None,count=None):
       randomUtils.forwardSeed(count, engine=eng)
     self.randomEng = eng
 
+git
```
Comment (PaulTalbot-INL) [inline ravenframework/SupervisedLearning/ARMA.py:2531]: extraneous entry?

File: ravenframework/SupervisedLearning/ROMCollection.py [hunk 2]
```diff
@@ -903,6 +903,7 @@ def evaluate(self, edict):
         result, weights = self._createNDEvaluation(edict)
       clusterStartIndex = 0 # what index does this cluster start on in the truncated signal?
       globalLocalPicker = []
+      nRoms = len(self._roms)
```
Comment (PaulTalbot-INL) [inline ravenframework/SupervisedLearning/ROMCollection.py:906]: does this variable get used later?

File: tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/Clustered_A_0.csv [hunk 3]
```diff
@@ -0,0 +1,1001 @@
+pivot,signal0,signal1
+0.000000000000000000e+00,-1.023538297423564752e-01,-1.475257076097549414e+00
+1.000000000000000056e-01,1.882947541575958894e-01,2.134044184248994558e+00
+2.000000000000000111e-01,-1.655719804495711101e-01,-2.284535550657079028e+00
+2.999999999999999889e-01,-3.606511423985305775e-01,-1.602966622035592703e+00
+4.000000000000000222e-01,8.025647150520918149e-01,5.933880922474279451e-01
+5.000000000000000000e-01,1.097985274012541090e+00,2.044384797439171564e-01
+5.999999999999999778e-01,5.954465753781293946e-01,6.307932507459796945e-01
+6.999999999999999556e-01,4.385963641041659145e-01,-5.825439638595943936e-01
+8.000000000000000444e-01,6.038094658580023255e-01,3.474744035754482319e+00
```
Comment (PaulTalbot-INL) [inline tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/Clustered_A_0.csv:10]: is there a reasonable way we could test this with a shorter data set? We typically try to keep regression test examples as lean as possible (but not leaner).

File: ravenframework/SupervisedLearning/ARMA.py (hunk 1 above)
Comment (j-bryan) [reply ravenframework/SupervisedLearning/ARMA.py:2531]: Yes, I saw that after I pushed. I thought I had pushed a fix earlier but maybe that's not showing up yet. I'll get that removed if it's still in the PR changes.

File: ravenframework/TSA/ARMA.py [hunk 4]
```diff
@@ -177,10 +180,10 @@ def characterize(self, signal, pivot, targets, settings):
       selCov = r.dot(q).dot(r.T)
       initCov = sp.linalg.solve_discrete_lyapunov(smoother['transition',:,:,0], selCov)
       initDist = {'mean': initMean, 'cov': initCov}
-      params[target]['arma'] = {'const': res.params[0], # exog/intercept/constant
-                                'ar': res.arparams,     # AR
-                                'ma': res.maparams,     # MA
-                                'var': res.params[-1],  # variance
+      params[target]['arma'] = {'const': res.params[res.param_names.index('const')], # exog/intercept/constant
+                                'ar': -res.polynomial_ar[1:],     # AR
```
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply ravenframework/TSA/ARMA.py:184]: sounds good to me.

File: tests/framework/ROM/TimeSeries/SyntheticHistory/clustered.xml [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    <created>2021-02-11</created>
+    <classesTested>SupervisedLearning.SyntheticHistory,TSA.Fourier,TSA.ARMA</classesTested>
+    <description>
+      Tests clustering for the SyntheticHistory ROM with Fourier and ARMA steps.
+    </description>
+  </TestInfo>
+
+  <RunInfo>
+    <WorkingDir>Clustered</WorkingDir>
+    <Sequence>read, train, print, sample, plot</Sequence>
```
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/SyntheticHistory/clustered.xml:15]: if they're in the other ones, we can leave these and choose to wholesale keep or remove them as we go. I tend to oscillate between choosing "fast" and "useful" for regression tests.

File: tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/Clustered_A_0.csv (hunk 3 above)
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/Clustered_A_0.csv:10]: That's a good point, I forgot this represents all the data that would then be segmented. What is the rough runtime for this test?

File: ravenframework/SupervisedLearning/ROMCollection.py (hunk 2 above)
Comment (PaulTalbot-INL) [reply ravenframework/SupervisedLearning/ROMCollection.py:906]: I don't see it showing up in the rest of the method, so I think we can remove it. I'm sure I had something planned for it that never came to fruition.

File: tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/Clustered_A_0.csv (hunk 3 above)
Comment (j-bryan) [reply tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/Clustered_A_0.csv:10]: About 5 seconds on my somewhat dated macbook air.

Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/SyntheticHistory/TrainingData/Clustered_A_0.csv:10]: Okay, that's trending towards longer, but not bad given the problem. I think it's fine.

**Review by PaulTalbot-INL — APPROVED** [review]

**Pull request merged**